pythonwhat/parsing.py

```diff
@@ -455,6 +449,86 @@ def visit_Lambda(self, node):
                 "body": node.body
             })
 
+
+class CompParser(Parser):
+    def __init__(self):
+        self.comps = []
+
+    def visit_Assign(self, node):
+        self.visit(node.value)
+
+    def visit_AugAssign(self, node):
+        self.visit(node.value)
+
+    def visit_Try(self, node):
+        self.visit_each(node.body)
+        self.visit_each(node.finalbody)
+
+    def visit_TryFinally(self, node):
+        self.visit_each(node.body)
+        self.visit_each(node.finalbody)
+
+class ListCompParser(CompParser):
+    """Find list comprehensions
+
+    A parser which inherits from the CompParser to find list comprehensions.
+    """
+    def visit_ListComp(self, node):
+        # ONLY single generators supported for now
+        target = node.generators[0].target
+        self.comps.append({
+                "list_comp": node,
+                "body": node.elt,
+                "target": target,
+                "target_vars": get_target_vars(target),
+                "iter": node.generators[0].iter,
+                "ifs": node.generators[0].ifs
+            })
+
+class DictCompParser(CompParser):
+    """Find dictionary comprehensions
+
+    A parser which inherits from the CompParser to find dict comprehensions.
+    """
+    def visit_DictComp(self, node):
+        # ONLY single generators supported for now
+        target = node.generators[0].target
+        self.comps.append({
+                "list_comp": node,
+                "key": node.key,
+                "value": node.value,
+                "target": target,
+                "target_vars": get_target_vars(target),
+                "iter": node.generators[0].iter,
+                "ifs": node.generators[0].ifs
+            })
+
+class GeneratorExpParser(CompParser):
+    """Find generator expressions
+
+    A parser which inherits from the CompParser to find generator expressions.
+    """
+
+    def visit_GeneratorExp(self, node):
+        # ONLY single generators supported for now
+        target = node.generators[0].target
+        self.comps.append({
+                "list_comp": node,
+                "body": node.elt,
+                "target": target,
+                "target_vars": get_target_vars(target),
+                "iter": node.generators[0].iter,
+                "ifs": node.generators[0].ifs
+            })
+
+def get_target_vars(target):
```

Comment:
Well in this case for example, you're using `get_target_vars()` in `ForParser` and `CompParser`, so the closest common superclass is `Parser`. So in this case I'd make `get_target_vars(target)` a static method in parser. This works as well, it's just less OO.